Fix exception when using base types of extension-based types from C#

mihe opened the pull request:
Fixes #74801.

This PR allows for GDExtension-based classes to be interacted with from C# so long as it's done through one of the engine-based base classes.

For example, without this fix the following code will cause a `NullReferenceException` when using a GDExtension-based physics server, because in such cases the `state` parameter must be backed by an extension class inheriting from `PhysicsDirectBodyState3DExtension`, which causes the C# integration to look for that extension class within the `GodotSharp` assembly, which it won't find.

```cs
public partial class MyRigidBody : RigidBody3D
{
    public override void _IntegrateForces(PhysicsDirectBodyState3D state)
    {
    }
}
```

This fix solves this (somewhat crudely) by traversing the class hierarchy when looking for the appropriate `ClassInfo`, and skipping them so long as it's a GDExtension class or its name happens to be either `PhysicsServer2DExtension` or `PhysicsServer3DExtension`. The exceptions for the physics server classes have to be made due to a workaround introduced in #59286 which explicitly filters them out when generating the C# bindings.

Note that this fix doesn't do anything to solve the issue of actually exposing the GDExtension class itself to C#. If anything, this fix would probably get in the way of solving that problem, but I figured this could be a stopgap solution.

EDIT: MRP can be found [here](https://github.com/godotengine/godot/files/11725673/PR75955_MRP.zip).

mihe committed:
Fix crash when using base types of extension-based types from C#

RedworkDE commented on modules/mono/csharp_script.cpp:
```diff
@@ -1203,7 +1206,7 @@ bool CSharpLanguage::setup_csharp_script_binding(CSharpScriptBinding &r_script_b
 	// ¯\_(ツ)_/¯
 	const ClassDB::ClassInfo *classinfo = ClassDB::classes.getptr(type_name);
-	while (classinfo && !classinfo->exposed) {
+	while (classinfo && (!classinfo->exposed || classinfo->gdextension || ignored_types.has(classinfo->name))) {
```
1. I would add a comment calling this out as a workaround / hack with a link to the issue / PR, to make it easier to find when eventually fixing this properly
2. Pretty sure that singletons have to be ignored as well, or the ignored types list is pointless (it would just try to instanciate static classes instead and still crash anyways)

RedworkDE replied:
Scenario: A method returns an extension implemented physics server singleton instance.

When this method is called from C# a wrapper object has to be created.
This code determines what type the wrapper should be, the actual extension class is disqualified, next the server extension is explicitly ignored, next the physics server type itself is chosen.
The wrapper then needs to be instanciated: while the managed type that corresponds to the native type can be found, it cannot be instanciated because it is a static class and static classes cannot have instances, thus error and probably a crash.

Because `Engine.GetSingleton` exists, there has to be some existing handling of this case, but I am not sure how exactly it works and I am not at my PC to check.
Depending on how exactly this works, ignoring the extensions here is either pointless, because singleton instances are alredy taken care of elsewhere, or singleton types need to be ignored here as well so that wrappers of type `Object` or `RefCounted` are properly created.

raulsntos commented:
I think this is a good temporary solution until the .NET module implements proper support for GDExtension classes.

I haven't tested this very thoroughly, but it allows getting a GDExtension node with `GetNode` and it's possible to access its members with the `Call`/`Get`/`Set` methods in `GodotObject`.

YuriSizov commented:
@raulsntos Would you mind giving this another look and an approval if you find it a good temp solution?

mihe commented:
I went ahead and rebased against latest master (593d5ca29f32d07a8bf9f3bae5b33e67d0f4218f), just in case. I also updated the comments slightly.

I also took the time to create a one-off build of [my GDExtension](https://github.com/godot-jolt/godot-jolt) that works with latest Godot master, which as of writing this is 593d5ca29f32d07a8bf9f3bae5b33e67d0f4218f, and have included that in an MRP that you'll find below. I assume it'll be forward-compatible for the foreseeable future, but that's the specific commit I targeted if you need it.

With this MRP you'll immediately run into the errors that prompted this PR to begin with, like not being able to create the instance binding for my `JoltPhysicsServer3D`, and when applying the changes in this PR you'll see that the errors go away.

I'll be honest and say that I have not tested this extensively, nor with any larger project, but the changes seem straight-forward enough that I don't see why they wouldn't scale.

Anyway, here's the MRP: [PR75955_MRP.zip](https://github.com/godotengine/godot/files/11725673/PR75955_MRP.zip)

Let me know if there's anything else I can do to help move things along.

raulsntos reviewed (approved):
Thanks for the MRP, it was very helpful. I also tested it a bit with [Godot Rust](https://github.com/godot-rust/gdext), this PR will also fix an issue they were having (https://github.com/godot-rust/gdext/issues/166).

The PR looks good to me so I'm approving.

akien-mga commented:
Thanks!
